src/tally/TrackLengthMeshTally.cpp

```diff
@@ -0,0 +1,721 @@
+    return;
+  }
+
+  double weight = event.get_score_multiplier(input_data.multiplier_id);
+
+  std::vector<double> intersections;     // vector of distance to triangular facet intersections
+  std::vector< EntityHandle > triangles; // vector of entityhandles that belong to the triangles hit
+  //  std::vector<ray_data> hit_information; // vector of reformated ray triangle intersections
+
+  // get all ray-triangle intersections along the ray
+  ErrorCode rval = get_all_intersections(event.position, event.direction, event.track_length, triangles, intersections);
+  if (rval != MB_SUCCESS) {
+    std::cout << "we have a problem finding intersections" << std::endl;
+    exit(1);
+  }
+
+  EntityHandle tet; // tet
+  if (intersections.size() == 0)
+    // ray is so short it either does not intersect a triangular face, or it inside the mesh
+    // but can't reach
+  {
+    // tet = point_in_which_tet(event.position);
+    tet = point_in_which_tet(event.position);
+    // if tet value is greater than 0 then in a tet, otherwise not
+    if (tet == 0) {
+      return;
+    } else {
+      // determine tracklength to return
+      add_score_to_mesh_tally(tet, weight, event.track_length, ebin);
+      //    found_crossing = true;
+      return;
+    }
+  }
+
+  // sort the intersection data
+  sort_intersection_data(intersections, triangles);
+
+  // compute the tracklengths
+  compute_tracklengths(event, ebin, weight, intersections, triangles);
+
+  return;
+}
+
+
+//---------------------------------------------------------------------------//
+// This may not need to be overridden, depending on whether conformality
+void TrackLengthMeshTally::end_history() {
+  MeshTally::end_history();
+  if (!conformality.empty()) {
+    last_cell = -1;
+  }
+}
+
+//---------------------------------------------------------------------------//
+void TrackLengthMeshTally::write_data(double num_histories) {
+  ErrorCode rval;
+
+  Range all_tets;
+  rval = mb->get_entities_by_dimension(tally_mesh_set, 3, all_tets);
+  if (rval != MB_SUCCESS) {
+    std::cout << "Failed to get 3d entities" << std::endl;
+    exit(1);
+  }
+  assert(rval == MB_SUCCESS);
+
+
+  for (Range::const_iterator i = all_tets.begin(); i != all_tets.end(); ++i) {
+    EntityHandle t = *i;
+
+    CartVect v[4];
+
+    std::vector<EntityHandle> vtx;
+    mb->get_connectivity(&t, 1, vtx);
+    assert(vtx.size() == 4);
+
+    int k = 0;
+    for (std::vector<EntityHandle>::iterator j = vtx.begin(); j != vtx.end(); ++j) {
+      EntityHandle vertex = *j;
+      mb->get_coords(&vertex, 1, v[k++].array());
+    }
+
+    double volume = tet_volume(v[0], v[1], v[2], v[3]);
+    unsigned int tet_index = get_entity_index(t);
+
+    unsigned int num_ebins = data->get_num_energy_bins();
+
+    // if there is a total, dont do anything with it
+    if (data->has_total_energy_bin())
+      num_ebins--;
+
+    double tally_vect[num_ebins];
+    double error_vect[num_ebins];
```

Comment:
Variable length arrays are not supported in C++, though since they are in newer versions of C many C++ compilers support them. I'll suggest the changes required to use vectors instead here, but it's not necessarily required to use them. For these two lines, change them to:
```c++
  std::vector<double> tally_vect(num_ebins);
  std::vector<double> error_vect(num_ebins);
```
One more set of changes (two lines) below.